### PR TITLE
Add FreeBSD to ifdef

### PR DESCRIPTION
**Pull request by pkubaj**
FreeBSD has this function as well.

### CODE REVIEW

**Comment by janisozaur** [comment]
I will likely have to reject this, as it seems you and @Gymnasiast misunderstood the purpose of this piece of code.

`strcasestr` is a utility function that is provided by many libc implementations, including glibc, but not all of them. In cases where it is _not_ provided, we ship our own implementation, see https://github.com/OpenRCT2/OpenRCT2/blob/e7af3290c0fe96036ffdfea79f15c7e9106223fa/src/openrct2/util/Util.cpp#L486-L523 guarded by the very same guards as the header.

@pkubaj your comment explicitly states that this function is provided by whatever libc FreeBSD ships with, so I see no point in this change, especially when you have not modified the aforementioned implementation guards, but as I might be overlooking something, I'm going to wait for your response.

**Comment by Gymnasiast** [comment]
@janisozaur I think you misread the if statement. It's `not(a or (b and c) or d)`. Basically, it uses one not operator for everything. I nearly misread it myself, but it is correct.

**Comment by janisozaur** [comment]
@Gymnasiast doesn't really matter, as the definition is still using non-updated guards.

**Comment by janisozaur** [comment]
@pkubaj @jbeich any comments?

**Comment by jbeich** [comment]
`strcasestr()` also exists on [DragonFly](https://www.dragonflybsd.org/cgi/web-man?command=strcasestr&section=3), [NetBSD](http://netbsd.gw.com/cgi-bin/man-cgi?strcasestr+3+NetBSD-current), [OpenBSD](https://man.openbsd.org/strcasestr.3) and Solaris ([IllumOS](https://illumos.org/man/3C/strcasestr), [Oracle](https://docs.oracle.com/cd/E36784_01/html/E36874/strcasestr-3c.html)). Each OS hides the prototype under different guard. I don't use OpenRCT2 but why not autodetect?
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2b94d9a14..39f6c3541 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -10,6 +10,7 @@ if (NOT MSVC)
     include(FindPkgConfig)
 endif ()
 include(CheckCXXCompilerFlag)
+include(CheckFunctionExists)
 include(GNUInstallDirs)
 
 set(CMAKE_CXX_STANDARD 17)
@@ -149,6 +150,10 @@ else ()
 endif ()
 
 # Defines
+check_function_exists(strcasestr HAVE_STRCASESTR)
+if (HAVE_STRCASESTR)
+    add_definitions(-DHAVE_STRCASESTR)
+endif()
 if (USE_MMAP)
     add_definitions(-DUSE_MMAP)
 endif ()
diff --git a/src/openrct2/util/Util.cpp b/src/openrct2/util/Util.cpp
index 26ba239cd..269512af2 100644
--- a/src/openrct2/util/Util.cpp
+++ b/src/openrct2/util/Util.cpp
@@ -483,7 +483,7 @@ char* safe_strtrimleft(char* destination, const char* source, size_t size)
     return safe_strcpy(destination, source, size);
 }
 
-#if !(defined(_GNU_SOURCE) || (defined(__DARWIN_C_LEVEL) && __DARWIN_C_LEVEL >= 200809L))
+#ifndef HAVE_STRCASESTR
 char* strcasestr(const char* haystack, const char* needle)
 {
     const char* p1 = haystack;
diff --git a/src/openrct2/util/Util.h b/src/openrct2/util/Util.h
index bf4dc23f4..5d2ecd291 100644
--- a/src/openrct2/util/Util.h
+++ b/src/openrct2/util/Util.h
@@ -44,7 +44,7 @@ char* safe_strcpy(char* destination, const char* source, size_t num);
 char* safe_strcat(char* destination, const char* source, size_t size);
 char* safe_strcat_path(char* destination, const char* source, size_t size);
 char* safe_strtrimleft(char* destination, const char* source, size_t size);
-#if !(defined(_GNU_SOURCE) || (defined(__DARWIN_C_LEVEL) && __DARWIN_C_LEVEL >= 200809L))
+#ifndef HAVE_STRCASESTR
 char* strcasestr(const char* haystack, const char* needle);
 #endif
 
```

**Comment by jbeich** [comment]
Doh, FreeBSD cannot use bundled version:
```
$ c++ -v
FreeBSD clang version 7.0.0 (branches/release_70 340910) (based on LLVM 7.0.0)
Target: x86_64-unknown-freebsd12.0
Thread model: posix
InstalledDir: /usr/bin

$ cmake -GNinja ../OpenRCT2
$ ninja
[...]
In file included from ../OpenRCT2/src/openrct2/util/Util.cpp:13:
In file included from ../OpenRCT2/src/openrct2/util/../core/Guard.hpp:12:
In file included from /usr/include/c++/v1/memory:658:
In file included from /usr/include/c++/v1/utility:204:
In file included from /usr/include/c++/v1/cstring:61:
In file included from /usr/include/c++/v1/string.h:61:
/usr/include/string.h:75:7: error: declaration of 'strcasestr' has a different language linkage
char    *strcasestr(const char *, const char *) __pure;
         ^
../OpenRCT2/src/openrct2/util/Util.h:48:7: note: previous declaration is here
char* strcasestr(const char* haystack, const char* needle);
      ^
1 error generated.
```

**Comment by janisozaur** [comment]
Sooo… is the version on current `develop` correct?

**Comment by Gymnasiast** [comment]
Not necessarily, the error that @jbeich has seems to be caused by ifdef'ing out our own definition, but not our declaration. Both should be ifdef'ed out, as you (janisozaur) already pointed out earlier.

File: src/openrct2/util/Util.h [hunk 1]
```diff
@@ -44,7 +44,7 @@ char* safe_strcpy(char* destination, const char* source, size_t num);
 char* safe_strcat(char* destination, const char* source, size_t size);
 char* safe_strcat_path(char* destination, const char* source, size_t size);
 char* safe_strtrimleft(char* destination, const char* source, size_t size);
-#if !(defined(_GNU_SOURCE) || (defined(__DARWIN_C_LEVEL) && __DARWIN_C_LEVEL >= 200809L))
+#if !(defined(_GNU_SOURCE) || (defined(__DARWIN_C_LEVEL) && __DARWIN_C_LEVEL >= 200809L) || defined(__FreeBSD__))
```
Comment (jbeich) [inline src/openrct2/util/Util.h]: To be consistent `__DARWIN_C_LEVEL` abomination replace `__FreeBSD__` with `__BSD_VISIBLE`. That would make the conditional respect `-D_XOPEN_SOURCE=<...>` or `-D_POSIX_C_SOURCE=<...>`.

**Comment by jbeich** [comment]
@Gymnasiast writes:
> the error that @jbeich has seems to be caused by ifdef'ing out our own definition, but not our declaration

No, 03cd16220 tree with no patches is busted. I just didn't notice initially. 4a379c6 having minor inconsistency between declaration (system) and definition (bundled) doesn't break build because the symbol has same ABI.

**Comment by janisozaur** [comment]
I've implemented the simpler solution in https://github.com/OpenRCT2/OpenRCT2/pull/7942, can you please test it?

**Comment by jbeich** [comment]
I confirm, #7942 applied to `develop` it builds fine on FreeBSD.

**Comment by janisozaur** [comment]
Thanks (especially for listing out the systems in https://github.com/OpenRCT2/OpenRCT2/pull/7942#issuecomment-417449993), I will apply once CIs are green.

**Pull request closed without merging**